cli_tools/common/distro/distro.go

```diff
@@ -106,25 +132,38 @@ func FromComponents(distro string, major string, minor string) (r Release, e err
 //
 // https://cloud.google.com/sdk/gcloud/reference/compute/images/import#--os
 func FromGcloudOSArgument(osFlagValue string) (r Release, e error) {
-	os := strings.ToLower(osFlagValue)
-	if strings.HasSuffix(os, "-byol") {
-		os = strings.TrimSuffix(os, "-byol")
+	if osFlagValue == "windows-8-1-x64-byol" {
+		// windows-8-1-x64-byol is a legacy flag value, and it's the only value that
+		// includes an extra hyphen between its major and minor version. The non-legacy
+		// flag is windows-8-x64-byol.
+		osFlagValue = "windows-8-x64-byol"
+	}
+	match := osFlagExpression.FindStringSubmatch(strings.ToLower(osFlagValue))
+	if match == nil {
+		return r, fmt.Errorf("expected pattern of `distro-version`. Actual: `%s`", osFlagValue)
 	}
-	lastHyphenIndex := strings.LastIndex(os, "-")
-	if lastHyphenIndex < 0 || lastHyphenIndex == len(os) {
-		return r, fmt.Errorf("expected pattern of `distro-version`. Actual: `%s`", os)
+	components := make(map[string]string)
+	for i, name := range osFlagExpression.SubexpNames() {
+		if i != 0 && name != "" {
+			components[name] = match[i]
+		}
 	}
-	distro, version := os[:lastHyphenIndex], os[lastHyphenIndex+1:]
+	distro, version, arch := components["distro"], components["version"], components["arch"]
 
+	var major, minor string
 	if distro == ubuntu {
 		// In gcloud, major and minor are combined as MMmm, such as ubuntu-1804
 		if len(version) != 4 {
 			return r, fmt.Errorf("expected version with length four. Actual: `%s`", version)
 		}
 
-		return FromComponents("ubuntu", version[:2], version[2:])
+		major, minor = version[:2], version[2:]
+	} else if distro == windows && strings.HasSuffix(version, "r2") {
```

Comment:
"r2" is used in multiple places in multiple functions, extract as a constant

Reply:
done!